src/clients/coprocessor.rs

```diff
@@ -1,34 +1,186 @@
+    pub coprocessor: String,
+}
+
+impl Default for CoprocessorClient {
+    fn default() -> Self {
+        Self::new(Self::DEFAULT_COPROCESSOR.into())
+    }
 }
 
 impl CoprocessorClient {
+    /// The default co-processor public address.
+    pub const DEFAULT_COPROCESSOR: &str = "https://service.coprocessor.valence.zone";
+
+    /// The deployed domain prover circuit id.
+    pub const DOMAIN_CIRCUIT: &str =
+        "cf4d4c2f3bf4ad0114091ea8023ff2456d5572e68bbc4d4e91bfa8f4a6f5d502";
+
+    pub fn new(coprocessor: String) -> Self {
+        Self { coprocessor }
+    }
+
     /// Starts a client that connects to a localhost co-processor
     pub fn local() -> Self {
         Self {
-            client: Client::local(),
+            coprocessor: "http://127.0.0.1:37281".into(),
+        }
+    }
+
+    /// Computes the URI of the co-processor.
+    pub fn uri<P: AsRef<str>>(&self, path: P) -> String {
+        format!("{}/api/{}", self.coprocessor, path.as_ref(),)
+    }
+
+    /// Fetches a proof from the queue, returning if present.
+    pub async fn get_proof_from_storage<C: AsRef<str>, P: AsRef<str>>(
+        &self,
+        circuit: C,
+        path: P,
+    ) -> anyhow::Result<Option<Proof>> {
+        let uri = format!("registry/controller/{}/storage/fs", circuit.as_ref());
+        let uri = self.uri(uri);
+
+        let response = reqwest::Client::new()
+            .post(uri)
+            .json(&json!({
+                "path": path.as_ref()
+            }))
+            .send()
+            .await?
+            .json::<Value>()
+            .await?;
+
+        let data = match response.get("data") {
+            Some(d) => d,
+            _ => return Ok(None),
+        };
+
+        let data = data
+            .as_str()
+            .ok_or_else(|| anyhow::anyhow!("unexpected data format"))?;
+
+        if data.is_empty() {
+            return Ok(None);
+        }
+
+        let data = Base64::decode(data)?;
+        let data: Value = serde_json::from_slice(&data)?;
+
+        anyhow::ensure!(
+            data.get("success")
+                .and_then(Value::as_bool)
+                .unwrap_or(false),
+            "the proof was computed incorrectly"
+        );
+
+        let proof = data
+            .get("proof")
+            .and_then(Value::as_str)
+            .ok_or_else(|| anyhow::anyhow!("failed to get proof from response"))?;
+
+        let bytes = Base64::decode(proof)?;
+        let proof = Proof::unpack(&bytes)
+            .map_err(|e| anyhow::anyhow!("failed to unpack proof: {e}"))?
+            .1;
+
+        Ok(Some(proof))
+    }
+
+    async fn _prove(&self, circuit: &str, path: &str, frequency: u64) -> anyhow::Result<Proof> {
+        let frequency = Duration::from_millis(frequency);
+
+        loop {
+            if let Some(p) = self.get_proof_from_storage(circuit, path).await? {
+                return Ok(p);
+            }
+
+            time::sleep(frequency).await;
+        }
+    }
```

Comment:
right, I think this addresses my above point about the `Result<Option<Proof>>` response type :+1: